File: 3731-find-missing-elements/3731-find-missing-elements.java
class Solution {
    public List<Integer> findMissingElements(int[] nums) {
        int n = nums.length;
        // this is finding min and max in one pass
        int min = nums[0], max = nums[0];
        for (int i = 1; i < n; i++) {
            if (nums[i] < min) min = nums[i];
            if (nums[i] > max) max = nums[i];
        }

        // this is using a boolean array for O(1) lookup
        boolean[] present = new boolean[max - min + 1];
        for (int num : nums) {
            present[num - min] = true;
        }

        // this is collecting missing elements in sorted order
        List<Integer> result = new ArrayList<>();
        for (int i = 1; i < present.length - 1; i++) {
            if (!present[i]) {
                result.add(i + min); // this is converting back to actual value
            }
        }

        return result;
    }
}

File: 3731-find-missing-elements/3731-find-missing-elements.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var findMissingElements = function(nums) {
    const n = nums.length;
    if (n === 0) return [];

    // Find min and max in one pass
    let min = nums[0];
    let max = nums[0];
    for (let i = 1; i < n; i++) {
        if (nums[i] < min) min = nums[i];
        if (nums[i] > max) max = nums[i];
    }

    // Using a Uint8Array or standard array as a boolean buffer map
    const range = max - min + 1;
    const present = new Uint8Array(range);
    for (let i = 0; i < n; i++) {
        present[nums[i] - min] = 1;
    }

    // Collecting missing interior elements in sorted order
    const result = [];
    for (let i = 1; i < range - 1; i++) {
        if (present[i] === 0) {
            result.push(i + min); // Converting back to actual value
        }
    }

    return result;
};
